Demystifying api (#887)
arcus/education_modules: demystifying_apis/demystifying_apis.md

<!--
module_id: demystifying_api
author:   Joy Payton
email:    paytonk@chop.edu
version:  1.0.0
current_version_description: Initial version
module_type: standard
docs_version: 4.0.0
language: en
narrator: US English Female
mode: Textbook
title: Demystifying Application Programming Interfaces (APIs) 
comment:  Understand what an application programming interface (API) is and why APIs are useful!
long_description: An application programming interface (API) can allow you to work more easily with data sources and technical tools.  Learn more about what an API is, read about use cases, and understand how interaction with APIs can take place.  It's appropriate for brand new beginners.
estimated_time_in_minutes: 30

@pre_reqs
No particular skills or experience are required for this module.  
@end

@learning_objectives  
After completion of this module, learners will be able to:

- Define the term *application programming interface* (API)
- Explain why APIs can be useful to biomedical researchers
- Explain how to interact with an API

@end

good_first_module: true
collection: demystifying, infrastructure_and_technology
coding_required: false

@sets_you_up_for

> 

Author: I agree!  Adding in a new commit.

A better approach is to use a script that uses an API call. First of all, it's scripted, which means no manual steps to write up in a Word document or add to a GitHub repo or jot down on a sticky note. Also, the typical API has a standard interface that will change very little. API access may improve over time, adding new features, but it's very infrequent that an API will radically change and remove options, rendering your script unusable. The same half-dozen lines of code you use to access your data will almost always be stable for months or years, and if you do need to change it, you're only changing that small chunk of code, instead of a step-by-step document with words and images that describe a manual effort.

## Interacting with an API 

How do you interact with an Application Programming Interface?  Well, first, someone (like the IT department at the New York Times or the developers who work behind the scenes at REDCap) needs to have written an API, a set of rules that allows you, or your computer system, to interrogate their server, which is always listening for HTTP messages. 

> You'll use your computer to send a message to the server asking for whatever you need from the API.  Then the server will send back some combination of a return code (such as 200, good, or 404, not found) and, if things went well, the data you requested in your message.

Reviewer: ```suggestion
You'll use your computer to send a message to the server asking for whatever you need from the API.  Then the server will send back some combination of a return code (such as 200, "good", or 404, "not found") and, if things went well, the data you requested in your message.
```

Let's talk about each one in turn! 

### Using a URL in Your Browser

Some very simple API calls that don't require authentication or lots of information exchange can be handled just in a URL (the web address you can type into a browser window).  

<div class = "important">
<b style="color: rgb(var(--color-highlight));">Important note</b><br>

Not every API can use this method!  Generally, the use of a URL is restricted to requests that are approved for anonymous users (so, things the application is fine sharing with the general public without a login), and for simple requests with only a few parameters.  

</div>

Let's look at an example!

NCBI, the National Center for Biotechnology Information, has an API called "Entrez" that you can use for a number of purposes, including to get PubMed article IDs that meet your conditions.  They provide the following example on [their website](https://www.ncbi.nlm.nih.gov/books/NBK25500/):

> Example: Get the PubMed IDs (PMIDs) for articles about breast cancer published in Science in 2008
>
> `https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=pubmed&term=science[journal]+AND+breast+cancer+AND+2008[pdat]`

> Note that there are some terms added in the URL, parameters like `db` and `term`.  There are clearly some rules about how `term` is structured, and if we wanted to learn more about this API we could read their documentation to learn what parameters we could include in the URL.

Reviewer: ```suggestion
Note that there are some terms that are included in the URL, parameters like `db` and `term`.  There are clearly some rules about how `term` is structured, and if we wanted to learn more about this API we could read their documentation to learn what parameters we could include in the URL.
```

Try opening that URL (<lia-keep><a href=https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=pubmed&term=science%5Bjournal%5D+AND+breast+cancer+AND+2008%5Bpdat%5D, target="blank">https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=pubmed&term=science[journal]+AND+breast+cancer+AND+2008[pdat]</a></lia-keep>) in your browser and you'll get some text that's clearly formatted for computers, not for humans!  It will look something like this:

![A screen full of XML encoded text that includes tags like "IdList", "eSearchResult", "QueryTranslation", and more.](media/entrez_screenshot.png)

Sometimes you don't need to fully automate something, you're happy just to look at the results in your browser.  Maybe you're just doing a one-off search that you don't need to repeat, or you're testing things out.  Often, when you're learning how to use a new API, it can be helpful to do a browser test like this to make sure you understand the syntax of how to use the API and are getting the results you expect. 

<div class = "help">
<b style="color: rgb(var(--color-highlight));">Troubleshooting help</b><br>

> Entrez, like many APIs, has a limit to what you can do anonymously.  If you try to run this a few times in a row, you'll get an error message that starts with `{"error":"API rate limit exceeded"`...

Reviewer: ```suggestion
E-utilities, like many APIs, has a limit to what you can do anonymously.  If you try to run this a few times in a row, you'll get an error message that starts with `{"error":"API rate limit exceeded"`...
```

<div class = "important">
<b style="color: rgb(var(--color-highlight));">Important note</b><br>

Note that here, we obscured the authentication token here.  While the data in this REDCap database is public data, it's still not a good idea to share authentication details!

</div>

### Constructing an API Call in a Programming Language: API-Specific Packages

We said on the last page that sometimes, specific APIs are very popular, and people write software (like a special library or package) to streamline using them.  This is true of APIs like those belonging to PubMed, the US Census Bureau, REDCap, and many more.  People write libraries to streamline the use of these popular APIs so you don't have to write quite as many lines of code as you saw on the previous page.

For example, `PyCap` is a Python library written to streamline access to the REDCap API.  Because "get all the records from a given database" is such a frequently used API request, it got its own functionality in `PyCap`.  Compare the code below with the code on the previous page.  It's much shorter and less intimidating!

> ![Six lines of Python code which get all of the records from a REDCap database ](media/pycap.png)

Reviewer: ```suggestion
![Six lines of Python code which get all of the records from a REDCap database.](media/pycap.png)
```

</div>
***

## Additional Resources

A number of libraries have lists of APIs that are useful for researchers, along with links to how to get started with each API.  The lists below have a number of overlaps, but we're including them all because there are also differences from institution to institution!

* [Temple University](https://guides.temple.edu/APIs)
* [University of California, Berkeley](https://guides.lib.berkeley.edu/information-studies/apis)
* [University of Virginia](https://guides.lib.virginia.edu/c.php?g=515084&p=10046079)
* [University of California, San Diego](https://ucsd.libguides.com/data-statistics/apis)

Some APIs (particularly those linked to social media applications) have changing rules relating to their use, due to monetization of data and data privacy.  Read more in [Shifting landscapes of social media data for research](https://www.timeshighereducation.com/campus/shifting-landscapes-social-media-data-research).

> The United States Office of the National Coordinator for Health Information Technology (ONC) is driving an effort to develop and standardize APIs related to electronic health information.  Read more about [their project](https://www.healthit.gov/topic/scientific-initiatives/accelerating-apis-scientific-discovery)!

Reviewer: ```suggestion
If you're interested in reading more about APIs in the biomedical and healthcare spaces, [the United States Office of the National Coordinator for Health Information Technology (ONC) is driving an effort to develop and standardize APIs related to electronic health information](https://www.healthit.gov/topic/scientific-initiatives/accelerating-apis-scientific-discovery). 
```
This is just a suggestion to try to make the link text itself a bit more descriptive, but I don't have super strong feelings about it.
